### PR TITLE
Add a lock around acquiring a token

### LINKED ISSUE #24: Support locks when signing in a user
Imagine:

* A site is deployed and all caches are cleared
* The site is very successful, and gets a ton of traffic
* The single, shared, cached token expires
* All of those requests see an expired token, and try to fetch a new one at the same time.
* This makes site owners and thePlatform sad as everything goes down.

We can fix this by supporting a locking service on the UserSession class.

- [x] Find out if there's a similar interface we can rely on like the PSR logging and cache interfaces, but for a lock service.
- [x] If not, see if any library (symfony? doctrine?) has one we can use.
- [x] Find or implement an equivalent to the NullLogger for logging. Making this a hard requirement (versus a set method) will help ensure implementors think about this potential DDoS target.

### PR DESCRIPTION
**Pull request by deviantintegral**
This resolves #24.

### CODE REVIEW

**Comment by deviantintegral** [comment]
@juanolalla codeclimate is reasonably flagging the 5-parameter constructor. Looking at the class, I'm thinking of trying to address #28 by moving the authentication methods like signIn, and their dependencies, to the User class. For example, `$user->signIn(...)` would return a UserSession. Does that seem like a good path to you?

File: src/Service/IdentityManagement/UserSession.php [hunk 1]
```diff
@@ -443,4 +459,31 @@ private function outerPromise(PromiseInterface $inner): Promise
...
+     *
+     * @param int $duration (optional) The number of seconds that the sign-in token should be valid for.
+     *
+     * @return Token
+     */
+    protected function signInWithLock(int $duration = null): Token
+    {
+        $factory = new Factory($this->store);
+        $factory->setLogger($this->logger);
+        $lock = $factory->createLock($this->user->getUsername(), 10);
```
Comment (deviantintegral) [inline src/Service/IdentityManagement/UserSession.php]: This creates a lock that lasts for 10 seconds. Note that any lock prefixing would be handled by decorating the store class that is passed in. The lock will release once `$lock` goes out of scope.

**Comment by juanolalla** [comment]
@deviantintegral At first look it seems reasonable to move the authentication methods to User. Once we do that, we should ask ourselves what is the purpose of Client and UserSession then. I would like to see only one class implementing the Guzzle ClientInterface if we don't really need two. That could be Client I think. And having UserSession for the rest of purposes left in the workflow.

**Comment by deviantintegral** [comment]
@juanolalla , I like where this is going. Now of course `User` has many constructor parameters and I'd like to fix that.

I'd like to add documentation and one more class so that:

* `\Lullabot\Mpx\Client` is the "root" MPX client. It works with all types of requests, authenticated or anonymous.
* `\Lullabot\Mpx\User` goes back to just having a username and password.
* `\Lullabot\Mpx\Service\IdentityManagement\UserSession` is renamed to `AuthenticatedClient`, stays as is in the current PR. It exists because of the complex retry / promise logic we need.
* A new class, but named `UserSession`, is forked out from what's currently in User. It handles the the token fetch and caching logic.

Plus, I'd add docs to the readme and class headers detailing all of this. Thoughts?

**Comment by juanolalla** [comment]
Well that sounds really good! Since there are clear reasons to have two clients, I'm glad you're changing the name and separating the other logic to a new UserSession class. I like classes to have a specific purpose, so that means having more classes when you realize the current ones are doing too much.

**Pull request merged**